lib/AST/Decl.cpp

```diff
@@ -2009,6 +2010,20 @@ VarDecl *VarDecl::getDefinition(ASTContext &C) {
   return nullptr;
 }
 
+// Checked C bounds information
+
+bool VarDecl::hasBoundsExpr() const {
+  return BoundsExpr != nullptr;
+}
+
+Expr *VarDecl::getBoundsExpr() {
+  return BoundsExpr;
+}
+
+void VarDecl::setBoundsExpr(Expr *E) {
```

Comment:
clang is using a region-based allocation scheme for AST nodes.  They are allocated into a region of memory associated with an ASTContext (see include\clang\AST\ASTContext.h).    Memory is allocated into the region using a bump-pointer based allocation scheme.   With this sort of approach, memory can only be freed when the entire region is freed.  Deallocation for an individual allocation doesn't doesn't do anything (see deallocate in ASTContext.h).   This means that code doesn't need to free nodes when references to them are dropped.  

I looked at the setInit code.  From what I can tell, there's memory allocated outside the ASTContext for the "Evaluated" meber that holds the result of compile-time evaluating the expression.  The main thing that setInit is doing in this case os running the destructor for EvaluatedStmt.  The destructor is implicitly generated. It runs the destructor for the Evaluated member in EvaluatedStmt, which looks like it frees some memory.